In [3]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns
# import plotly.express as px  # pip install plotly needs to executed
# import plotly.graph_objects as go

In [4]:
# Importing the data into Pandas dataframes

df_reports = pd.read_csv('data/reports.csv')
df_reviews = pd.read_csv('data/reviews.csv')
df_users = pd.read_csv('data/users.csv')

In [5]:
df_users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 432 entries, 0 to 431
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   UserId     432 non-null    int64 
 1   CountryId  432 non-null    object
 2   Age        432 non-null    int64 
 3   GenderId   432 non-null    int64 
dtypes: int64(3), object(1)
memory usage: 13.6+ KB


In [6]:
df_reports.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   ReportId  30 non-null     int64 
 1   Name      30 non-null     object
dtypes: int64(1), object(1)
memory usage: 608.0+ bytes


In [7]:
df_reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1800 entries, 0 to 1799
Data columns (total 19 columns):
 #   Column                Non-Null Count  Dtype
---  ------                --------------  -----
 0   ReportId              1800 non-null   int64
 1   ReviewId              1800 non-null   int64
 2   UserId                1800 non-null   int64
 3   Appeal                1800 non-null   int64
 4   FamiliarityWithBrand  1800 non-null   int64
 5   Bold                  1800 non-null   int64
 6   Pure                  1800 non-null   int64
 7   Defiant               1800 non-null   int64
 8   Fun-loving            1800 non-null   int64
 9   Intense               1800 non-null   int64
 10  Joyful                1800 non-null   int64
 11  Majestic              1800 non-null   int64
 12  Peaceful              1800 non-null   int64
 13  Confident             1800 non-null   int64
 14  Simple                1800 non-null   int64
 15  Sophisticated         1800 non-null   int64
 16  Sponta

In [8]:
# Need to create a new dataframe with all values combined

# df = df_reviews.replace(df_reviews['ReportId'], df_reports['Name'])

In [9]:
insert = {'ReportId': {5253001: 'Aldi', 
                        5252081: 'Amazon Prime Originals',
                        5252082: 'American Express',
                        5252083: 'Apple Mac',
                        5249364: 'ASDA',
                        5249629: 'Autoglass',
                        5253003: 'Betway',
                        5249631: 'Checkatrade',
                        5253598: 'Coca Cola',
                        5249634: 'Currys PC World',
                        5252209: 'Disney',
                        5252087: 'Disney+',
                        5249365: 'EE',
                        5249367: 'GoCompare',
                        5249368: 'Haribo',
                        5249369: 'Heart',
                        5249370: 'JUST EAT',
                        5249229: 'Lidl',
                        5252091: "McDonald's",
                        5252092: 'Netflix',
                        5253008: 'Ocado',
                        5252095: 'Pepsi',
                        5252096: 'Playstation',
                        5249697: 'Premier League',
                        5249699: 'TalkTalk',
                        5249405: 'Tesco',
                        5252099: 'Visa',
                        5252100: 'Warner Brothers',
                        5252101: 'Xbox One',
                        5252102: 'YouTube Originals'}
                    }

In [10]:
df_reviews = df_reviews.replace(insert)

In [11]:
df_reviews.head()

ReportId  ReviewId   UserId  Appeal  FamiliarityWithBrand  Bold  Pure  \
0     Lidl  28422162  3323715       4                    10     5     7   
1     Lidl  28422163   236732       7                     7     7     6   
2     Lidl  28422174  3358962       7                     4     7     8   
3     Lidl  28422209  2653082       9                     5     9     5   
4     Lidl  28422217  3342457       8                     0     6     8   

   Defiant  Fun-loving  Intense  Joyful  Majestic  Peaceful  Confident  \
0        6           4        5       6         4         7          5   
1        7           8        8       7         7         7          6   
2        7           8        6       7         8         8          9   
3        3           4        6       6         3         9          5   
4        2           5        0       8         0         6          4   

   Simple  Sophisticated  Spontaneous  Technical  Warm  
0       8              6            7          5     6  
1       8              6            7          7     6  
2       8              7            8          7     7  
3       5              3            7          5     7  
4      10              7            0          0     0

### 1.1 Compute the average attribute score for each brand

In [12]:
df_reviews.groupby(['ReportId'])['Bold', 'Pure', 'Defiant', 'Fun-loving', 'Intense',
                                    'Joyful', 'Majestic', 'Peaceful', 'Confident', 'Simple', 
                                    'Sophisticated', 'Spontaneous', 'Technical', 'Warm'].aggregate('mean')

Bold      Pure   Defiant  Fun-loving   Intense  \
ReportId                                                                     
ASDA                    5.833333  6.800000  4.516667    7.200000  4.683333   
Aldi                    7.466667  6.800000  5.700000    7.800000  6.250000   
Amazon Prime Originals  6.133333  6.983333  4.700000    6.366667  5.150000   
American Express        6.433333  6.450000  4.800000    6.233333  4.933333   
Apple Mac               7.216667  5.900000  5.416667    4.266667  5.583333   
Autoglass               7.416667  6.816667  5.283333    6.500000  5.250000   
Betway                  7.766667  5.750000  6.266667    6.350000  7.216667   
Checkatrade             7.566667  6.333333  5.866667    6.916667  5.933333   
Coca Cola               7.016667  7.116667  5.066667    7.050000  5.300000   
Currys PC World         7.583333  6.416667  5.950000    6.533333  6.133333   
Disney                  7.516667  8.116667  6.050000    8.300000  6.600000   
Disney+                 6.383333  6.200000  5.150000    4.883333  5.366667   
EE                      5.716667  4.833333  5.533333    5.466667  6.066667   
GoCompare               7.983333  6.483333  6.083333    7.533333  6.100000   
Haribo                  7.000000  6.650000  5.950000    8.183333  5.583333   
Heart                   7.433333  5.083333  5.883333    6.566667  6.550000   
JUST EAT                7.350000  6.066667  6.016667    7.616667  5.850000   
Lidl                    5.350000  6.800000  4.400000    6.950000  3.650000   
McDonald's              6.350000  6.466667  5.583333    7.333333  5.333333   
Netflix                 8.216667  6.266667  6.533333    6.116667  7.316667   
Ocado                   6.433333  6.850000  5.100000    7.200000  5.200000   
Pepsi                   7.366667  5.200000  6.066667    6.833333  6.283333   
Playstation             5.100000  3.583333  4.716667    3.600000  5.033333   
Premier League          6.516667  5.883333  6.100000    5.766667  6.400000   
TalkTalk                6.866667  6.766667  6.150000    6.933333  6.050000   
Tesco                   6.433333  6.350000  5.183333    6.000000  5.616667   
Visa                    5.433333  5.316667  4.716667    5.450000  4.750000   
Warner Brothers         7.516667  7.450000  5.850000    7.333333  6.533333   
Xbox One                6.866667  5.716667  5.233333    5.616667  6.000000   
YouTube Originals       6.166667  6.316667  5.383333    5.350000  5.416667   

                          Joyful  Majestic  Peaceful  Confident    Simple  \
ReportId                                                                    
ASDA                    7.483333  5.516667  6.266667   6.550000  7.316667   
Aldi                    7.616667  6.416667  5.500000   7.816667  6.550000   
Amazon Prime Originals  6.600000  5.916667  6.683333   6.516667  7.350000   
American Express        6.366667  5.383333  5.900000   6.383333  7.133333   
Apple Mac               4.300000  5.316667  5.683333   6.600000  7.816667   
Autoglass               6.983333  5.966667  6.266667   7.800000  7.966667   
Betway                  5.883333  5.866667  4.483333   7.516667  5.300000   
Checkatrade             6.866667  4.800000  5.283333   7.900000  7.766667   
Coca Cola               7.116667  5.900000  6.316667   7.516667  7.300000   
Currys PC World         6.600000  6.066667  5.983333   7.433333  6.716667   
Disney                  8.450000  8.316667  7.733333   8.033333  5.350000   
Disney+                 4.966667  5.483333  6.333333   6.183333  7.550000   
EE                      5.166667  4.783333  4.850000   5.850000  5.166667   
GoCompare               7.500000  7.066667  5.116667   8.183333  6.416667   
Haribo                  8.500000  5.583333  5.683333   7.066667  6.433333   
Heart                   6.683333  5.266667  4.750000   7.750000  6.300000   
JUST EAT                7.466667  6.016667  5.883333   7.733333  7.000000   
Lidl                    7.400000  5.550000  6.666667   6.366667  8.0

### 1.2 Turn the attribute scores into Z-scores.

In [13]:
from scipy.stats import zscore

In [14]:
df_avg = df_reviews.groupby(['ReportId'])['Bold', 'Pure', 'Defiant', 'Fun-loving', 'Intense',
                                    'Joyful', 'Majestic', 'Peaceful', 'Confident', 'Simple', 
                                    'Sophisticated', 'Spontaneous', 'Technical', 'Warm'].aggregate('mean')

df_avg

Bold      Pure   Defiant  Fun-loving   Intense  \
ReportId                                                                     
ASDA                    5.833333  6.800000  4.516667    7.200000  4.683333   
Aldi                    7.466667  6.800000  5.700000    7.800000  6.250000   
Amazon Prime Originals  6.133333  6.983333  4.700000    6.366667  5.150000   
American Express        6.433333  6.450000  4.800000    6.233333  4.933333   
Apple Mac               7.216667  5.900000  5.416667    4.266667  5.583333   
Autoglass               7.416667  6.816667  5.283333    6.500000  5.250000   
Betway                  7.766667  5.750000  6.266667    6.350000  7.216667   
Checkatrade             7.566667  6.333333  5.866667    6.916667  5.933333   
Coca Cola               7.016667  7.116667  5.066667    7.050000  5.300000   
Currys PC World         7.583333  6.416667  5.950000    6.533333  6.133333   
Disney                  7.516667  8.116667  6.050000    8.300000  6.600000   
Disney+                 6.383333  6.200000  5.150000    4.883333  5.366667   
EE                      5.716667  4.833333  5.533333    5.466667  6.066667   
GoCompare               7.983333  6.483333  6.083333    7.533333  6.100000   
Haribo                  7.000000  6.650000  5.950000    8.183333  5.583333   
Heart                   7.433333  5.083333  5.883333    6.566667  6.550000   
JUST EAT                7.350000  6.066667  6.016667    7.616667  5.850000   
Lidl                    5.350000  6.800000  4.400000    6.950000  3.650000   
McDonald's              6.350000  6.466667  5.583333    7.333333  5.333333   
Netflix                 8.216667  6.266667  6.533333    6.116667  7.316667   
Ocado                   6.433333  6.850000  5.100000    7.200000  5.200000   
Pepsi                   7.366667  5.200000  6.066667    6.833333  6.283333   
Playstation             5.100000  3.583333  4.716667    3.600000  5.033333   
Premier League          6.516667  5.883333  6.100000    5.766667  6.400000   
TalkTalk                6.866667  6.766667  6.150000    6.933333  6.050000   
Tesco                   6.433333  6.350000  5.183333    6.000000  5.616667   
Visa                    5.433333  5.316667  4.716667    5.450000  4.750000   
Warner Brothers         7.516667  7.450000  5.850000    7.333333  6.533333   
Xbox One                6.866667  5.716667  5.233333    5.616667  6.000000   
YouTube Originals       6.166667  6.316667  5.383333    5.350000  5.416667   

                          Joyful  Majestic  Peaceful  Confident    Simple  \
ReportId                                                                    
ASDA                    7.483333  5.516667  6.266667   6.550000  7.316667   
Aldi                    7.616667  6.416667  5.500000   7.816667  6.550000   
Amazon Prime Originals  6.600000  5.916667  6.683333   6.516667  7.350000   
American Express        6.366667  5.383333  5.900000   6.383333  7.133333   
Apple Mac               4.300000  5.316667  5.683333   6.600000  7.816667   
Autoglass               6.983333  5.966667  6.266667   7.800000  7.966667   
Betway                  5.883333  5.866667  4.483333   7.516667  5.300000   
Checkatrade             6.866667  4.800000  5.283333   7.900000  7.766667   
Coca Cola               7.116667  5.900000  6.316667   7.516667  7.300000   
Currys PC World         6.600000  6.066667  5.983333   7.433333  6.716667   
Disney                  8.450000  8.316667  7.733333   8.033333  5.350000   
Disney+                 4.966667  5.483333  6.333333   6.183333  7.550000   
EE                      5.166667  4.783333  4.850000   5.850000  5.166667   
GoCompare               7.500000  7.066667  5.116667   8.183333  6.416667   
Haribo                  8.500000  5.583333  5.683333   7.066667  6.433333   
Heart                   6.683333  5.266667  4.750000   7.750000  6.300000   
JUST EAT                7.466667  6.016667  5.883333   7.733333  7.000000   
Lidl                    7.400000  5.550000  6.666667   6.366667  8.0

In [15]:
df_z = df_avg.transform(lambda x: zscore(x))
df_z

Bold      Pure   Defiant  Fun-loving   Intense  \
ReportId                                                                     
ASDA                   -1.221420  0.636269 -1.737976    0.673706 -1.382636   
Aldi                    0.811975  0.636269  0.335911    1.231257  0.671649   
Amazon Prime Originals -0.847940  0.851843 -1.416670   -0.100669 -0.770721   
American Express       -0.474459  0.224719 -1.241412   -0.224569 -1.054824   
Apple Mac               0.500741 -0.422002 -0.160653   -2.052094 -0.202515   
Autoglass               0.749728  0.655867 -0.394331    0.023231 -0.639597   
Betway                  1.185456 -0.598381  1.329041   -0.116156  1.939187   
Checkatrade             0.936468  0.087536  0.628008    0.410419  0.256421   
Coca Cola               0.251754  1.008624 -0.774057    0.534319 -0.574034   
Currys PC World         0.957217  0.185524  0.774057    0.054206  0.518671   
Disney                  0.874222  2.184481  0.949315    1.695882  1.130585   
Disney+                -0.536706 -0.069245 -0.628008   -1.479057 -0.486618   
EE                     -1.366663 -1.676249  0.043815   -0.936994  0.431254   
GoCompare               1.455192  0.263915  1.007734    0.983457  0.474962   
Haribo                  0.231005  0.459891  0.774057    1.587469 -0.202515   
Heart                   0.770477 -1.382285  0.657218    0.085181  1.065023   
JUST EAT                0.666732 -0.226026  0.890896    1.060894  0.147151   
Lidl                   -1.823140  0.636269 -1.942444    0.441394 -2.737590   
McDonald's             -0.578204  0.244317  0.131444    0.797606 -0.530326   
Netflix                 1.745677  0.009146  1.796396   -0.332981  2.070311   
Ocado                  -0.474459  0.695062 -0.715637    0.673706 -0.705159   
Pepsi                   0.687481 -1.245102  0.978525    0.332981  0.715357   
Playstation            -2.134374 -3.146071 -1.387460   -2.671595 -0.923700   
Premier League         -0.370714 -0.441600  1.036944   -0.658219  0.868336   
TalkTalk                0.065013  0.597074  1.124573    0.425906  0.409400   
Tesco                  -0.474459  0.107134 -0.569589   -0.441394 -0.158806   
Visa                   -1.719395 -1.107919 -1.387460   -0.952482 -1.295219   
Warner Brothers         0.874222  1.400576  0.598799    0.797606  1.043169   
Xbox One                0.065013 -0.637576 -0.481960   -0.797606  0.343838   
YouTube Originals      -0.806442  0.067938 -0.219073   -1.045407 -0.421056   

                          Joyful  Majestic  Peaceful  Confident    Simple  \
ReportId                                                                    
ASDA                    0.850088 -0.292213  0.553387  -0.671333  0.663457   
Aldi                    0.970171  0.745911 -0.342944   1.046556 -0.279458   
Amazon Prime Originals  0.054538  0.169176  1.040524  -0.716540  0.704453   
American Express       -0.155608 -0.446009  0.124707  -0.897371  0.437977   
Apple Mac              -2.016895 -0.522907 -0.128604  -0.603521  1.278401   
Autoglass               0.399776  0.226849  0.553387   1.023952  1.462885   
Betway                 -0.590909  0.111502 -1.531558   0.639687 -1.816819   
Checkatrade             0.294704 -1.118867 -0.596255   1.159574  1.216907   
Coca Cola               0.519860  0.149951  0.611844   0.639687  0.642959   
Currys PC World         0.054538  0.342196  0.222134   0.526668 -0.074477   
Disney                  1.720690  2.937506  2.268108   1.340405 -1.755325   
Disney+                -1.416480 -0.330662  0.631329  -1.168616  0.950431   
EE                     -1.236355 -1.138091 -1.102877  -1.620692 -1.980805   
GoCompare               0.865098  1.495667 -0.791110   1.543839 -0.443443   
Haribo                  1.765721 -0.215315 -0.128604   0.029385 -0.422945   
Heart                   0.129590 -0.580580 -1.219790   0.956140 -0.586930   
JUST EAT                0.835078  0.284523  0.105222   0.933537  0.273992   
Lidl                    0.775036 -0.253764  1.021038  -0.919974  1.5

### 1.3 Prepare one or more graphs to show the difference in personality of the reports below

In [20]:
# Aldi, Lidl, ASDA, Tesco, Ocado

df_avg.transpose()

ax = sns.barplot(x="ReportId", y="Aldi", data=df_avg)

ValueError: Could not interpret input 'ReportId'

### 1.4 Analyse the difference in appeal between M and F respondents for the Disney brand

In [ ]:
# to do: compare correlations between familiarity and appeal for males and females separately
# first step: combining df_users into df_reviews 

### 1.5 Familiarity vs Appeal

In [21]:
df_reviews.corr(method='pearson')

ReviewId    UserId    Appeal  FamiliarityWithBrand  \
ReviewId              1.000000 -0.088728  0.026638              0.001160   
UserId               -0.088728  1.000000  0.027300              0.051219   
Appeal                0.026638  0.027300  1.000000              0.498920   
FamiliarityWithBrand  0.001160  0.051219  0.498920              1.000000   
Bold                 -0.001787  0.049856  0.572255              0.369757   
Pure                  0.011705  0.050246  0.614615              0.324275   
Defiant              -0.025195  0.035473  0.297986              0.251737   
Fun-loving           -0.098868  0.031881  0.602982              0.388293   
Intense               0.026782  0.050078  0.331706              0.242122   
Joyful               -0.117544  0.027081  0.596934              0.370522   
Majestic              0.015312  0.063696  0.549357              0.329930   
Peaceful             -0.006436 -0.003374  0.428413              0.243172   
Confident            -0.040819  0.015578  0.562438              0.349039   
Simple               -0.050715  0.002724  0.268780              0.179083   
Sophisticated         0.009843  0.003855  0.442373              0.288835   
Spontaneous          -0.151637 -0.033392  0.326052              0.254529   
Technical            -0.030918 -0.019295  0.364555              0.251534   
Warm                 -0.039884  0.011874  0.601397              0.406183   

                          Bold      Pure   Defiant  Fun-loving   Intense  \
ReviewId             -0.001787  0.011705 -0.025195   -0.098868  0.026782   
UserId                0.049856  0.050246  0.035473    0.031881  0.050078   
Appeal                0.572255  0.614615  0.297986    0.602982  0.331706   
FamiliarityWithBrand  0.369757  0.324275  0.251737    0.388293  0.242122   
Bold                  1.000000  0.407669  0.485397    0.425658  0.546579   
Pure                  0.407669  1.000000  0.276512    0.577305  0.277079   
Defiant               0.485397  0.276512  1.000000    0.322138  0.558879   
Fun-loving            0.425658  0.577305  0.322138    1.000000  0.281136   
Intense               0.546579  0.277079  0.558879    0.281136  1.000000   
Joyful                0.411788  0.599106  0.308345    0.773725  0.224070   
Majestic              0.478846  0.578489  0.433219    0.503565  0.482316   
Peaceful              0.240136  0.626745  0.324521    0.466543  0.175228   
Confident             0.628141  0.429775  0.384835    0.481288  0.468529   
Simple                0.198336  0.352157  0.157970    0.239846  0.078569   
Sophisticated         0.421958  0.444617  0.449639    0.363505  0.476979   
Spontaneous           0.421132  0.316687  0.470092    0.417304  0.464524   
Technical             0.448603  0.395920  0.482419    0.315603  0.515681   
Warm                  0.373576  0.646236  0.319547    0.667267  0.267494   

                        Joyful  Majestic  Peaceful  Confident    Simple  \
ReviewId             -0.117544  0.015312 -0.006436  -0.040819 -0.050715   
UserId                0.027081  0.063696 -0.003374   0.015578  0.002724   
Appeal                0.596934  0.549357  0.428413   0.562438  0.268780   
FamiliarityWithBrand  0.370522  0.329930  0.243172   0.349039  0.179083   
Bold                  0.411788  0.478846  0.240136   0.628141  0.198336   
Pure                  0.599106  0.578489  0.626745   0.429775  0.352157   
Defiant               0.308345  0.433219  0.324521   0.384835  0.157970   
Fun-loving            0.773725  0.503565  0.466543   0.481288  0.239846   
Intense               0.224070  0.482316  0.175228   0.468529  0.078569   
Joyful                1.000000  0.518053  0.527327   0.467345  0.269539   
Majestic              0.518053  1.000000  0.565033   0.453863  0.196525   
Peaceful              0.527327  0.565033  1.000000   0.255066  0.355459   
Confident             0.467345  0.453863  0.255066   1.000000  0.245572   
Simple                0.269539  0.196525  0.355459   0.245572  1.000000   
Sophist

There is a clear correlation of appeal and familiarity, coefficient 0.498920. 

## Assignment 2

In [22]:
# Using the data from assignment 1, build a model predicting the Appeal of a report based on its attribute scores

In [24]:
df_reviews.head(2)

ReportId  ReviewId   UserId  Appeal  FamiliarityWithBrand  Bold  Pure  \
0     Lidl  28422162  3323715       4                    10     5     7   
1     Lidl  28422163   236732       7                     7     7     6   

   Defiant  Fun-loving  Intense  Joyful  Majestic  Peaceful  Confident  \
0        6           4        5       6         4         7          5   
1        7           8        8       7         7         7          6   

   Simple  Sophisticated  Spontaneous  Technical  Warm  
0       8              6            7          5     6  
1       8              6            7          7     6

In [32]:
X = df_reviews.drop(['ReviewId', 'ReportId', 'UserId', 'FamiliarityWithBrand', 'Appeal'], axis=1)
y = df_reviews['Appeal']

In [33]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

In [34]:
print('X_train shape:', X_train.shape)
print('X_test shape:', X_test.shape)
print('y_train:', y_train.shape)
print('y_test:', y_test.shape)

X_train shape: (1440, 14)
X_test shape: (360, 14)
y_train: (1440,)
y_test: (360,)


In [35]:
from sklearn.linear_model import LinearRegression

In [37]:
linear_reg = LinearRegression()

linear_reg.fit(X_train, y_train)
y_pred = linear_reg.predict(X_test)

In [39]:
from sklearn.metrics import r2_score

r2_score(y_test, y_pred)
print(r2_score)

<function r2_score at 0x12f8dc1f0>


#### Random Forest

In [44]:
from sklearn.ensemble import RandomForestRegressor

In [45]:
rand_forst= RandomForestRegressor()

In [47]:
rand_forest_model = rand_forst.fit(X_train, y_train)
y_pred = rand_forst.predict(X_test)

In [48]:
import numpy as np
from sklearn import metrics

In [49]:
print('Mean Absolute Error (MAE):', metrics.mean_absolute_error(y_test, y_pred))
print('Mean Squared Error (MSE):', metrics.mean_squared_error(y_test, y_pred))
print('Root Mean Squared Error (RMSE):', metrics.mean_squared_error(y_test, y_pred, squared=False))

Mean Absolute Error (MAE): 0.9630833333333333
Mean Squared Error (MSE): 1.687508611111111
Root Mean Squared Error (RMSE): 1.2990414200906417
